### DISCUSSION TITLE
[9.x] Enable trait setUp/tearDown for tests traits.

### DISCUSSION THREAD

**Discussion by DarkGhostHunter** [laravel/framework#39287] [Ideas]
When dealing this testing, sometimes you may want to share initialization between many Test Classes. For example, you may want to set up a Post with an User and three comments.

```php
public function setUp(): void
{
    parent::setUp();

    $this->post = Post::factory()->hasComments(3)->createOne();
}

public function tearDown(): void
{
    parent:tearDown();

    $this->post = null;
}
```

Doing this across many Tests becomes cumbersome, specially if you need to mix additional setUp and tearDown procedures.

Instead, we can use the same technique of trait initialization and booting in Models, but in tests.

### Implementation

On the `setUpTraits()` method, we can setup the trait saving `setUp{trait}` and `tearDown{trait}` into a callback that later can be called by the application.

```php
protected function setUpTraits(): array
{
    foreach ($traits = class_uses_recursive(static::class)) as $trait) {
        if (method_exists($this, $name = 'setUp' . $trait)) {
            $this->afterApplicationCreated([$this, $name]);
        }

        if (method_exists($this, $name = 'tearDown' . $trait)) {
            $this->beforeApplicationDestroyed([$this, $name]);
        }
    }

    return $traits;
}
```

This would mean that traits like `RefreshDatabase`, `DatabaseMigrations`, `DatabaseTransactions`, `WithoutMiddleware`, `WithoutEvents` and `WithFaker` would change their booting methods, making it a breaking change for Laravel 9.x.

### Example

With the above, we can create a trait that creates a Post with an User and 3 Comments, keeping the implementation clean in the Test Class:

```php
class ExampleTest extends TestCase
{
    use CreatePost;

    public function test_post_sees_comments(): void
    {
        $this->get('post/' . $this->post->id)
            ->assertOk();
    }
}

trait CreatePost
{
    protected Post $post;

    protected function setUpCreatePost(): void
    {
        $this->post = Pots::factory()->hasComments(3)->createOne();
    }

    protected function setUpCreatePost(): void
    {
        $this->post = null;
    }
}
```

**Comment by crynobone** [comment]
I believe you already PR this and merged.
